sql/src/main/java/io/crate/executor/transport/ShardUpsertRequest.java

```diff
@@ -0,0 +1,323 @@
+        }
+
+        @Nullable
+        public String routing() {
+            return routing;
+        }
+
+        public long version() {
+            return version;
+        }
+
+        public int retryOnConflict() {
+            return version == Versions.MATCH_ANY ? Constants.UPDATE_RETRY_ON_CONFLICT : 0;
+        }
+
+        @Nullable
+        public Symbol[] assignments() {
+            return assignments;
+        }
+
+        @Nullable
+        public Object[] missingAssignments() {
+            return missingAssignments;
+        }
+
+        static Item readItem(StreamInput in, @Nullable Streamer[] streamers) throws IOException {
+            Item item = new Item(streamers);
+            item.readFrom(in);
+            return item;
+        }
+
+        @Override
+        public void readFrom(StreamInput in) throws IOException {
+            id = in.readString();
+            routing = in.readOptionalString();
+            int assignmentsSize = in.readVInt();
+            if (assignmentsSize > 0) {
+                assignments = new Symbol[assignmentsSize];
+                for (int i = 0; i < assignmentsSize; i++) {
+                    assignments[i] = Symbol.fromStream(in);
+                }
+            }
+            int missingAssignmentsSize = in.readVInt();
+            if (missingAssignmentsSize > 0) {
+                this.missingAssignments = new Object[missingAssignmentsSize];
+                for (int i = 0; i < missingAssignmentsSize; i++) {
+                    missingAssignments[i] = streamers[i].readValueFrom(in);
+                }
+            }
+
+            version = Versions.readVersion(in);
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeString(id);
+            out.writeOptionalString(routing);
+            if (assignments != null) {
+                out.writeVInt(assignments.length);
+                for (int i = 0; i < assignments.length; i++) {
+                    Symbol.toStream(assignments[i], out);
+                }
+            } else {
+                out.writeVInt(0);
+            }
+            // Stream References
+            if (missingAssignments != null) {
+                out.writeVInt(missingAssignments.length);
+                for (int i = 0; i < missingAssignments.length; i++) {
+                    streamers[i].writeValueTo(out, missingAssignments[i]);
+                }
+            } else {
+                out.writeVInt(0);
+            }
+
+            Versions.writeVersion(version, out);
+        }
+    }
+
+    private int shardId;
+    private List<Item> items;
+    private String[] assignmentsColumns;
+    private IntArrayList locations;
+    @Nullable
+    private Reference[] missingAssignmentsColumns;
+    @Nullable
+    private Streamer[] streamers;
+    private boolean continueOnError = false;
+
+    public ShardUpsertRequest() {
+    }
+
+    public ShardUpsertRequest(String index,
```

Comment:
seems like this ctor is only called from the other ctor further below.. so I guess it could be removed / changed to take shardId